src/test/java/org/jetbrains/kotlinx/lincheck/test/verifier/nesting_safe_recoverable_linearizability/set/NRLSet.java

```diff
@@ -0,0 +1,234 @@
+        for (int i = 0; i < threadsCount; i++) {
+            recoveryData.add(null);
+        }
+        checkPointer = new int[threadsCount];
+    }
+
+    @NotNull
+    private final List<Info> recoveryData;
+
+    @NotNull
+    private final int[] checkPointer;
+
+    @NotNull
+    private final AtomicReference<Node> head = new AtomicReference<>(null);
+
+    @NotNull
+    private PrevNextPair findPrevNext(@NotNull T value) {
+        start:
+        while (true) {
+            @Nullable Node previous = null;
+            @Nullable Node current = head.get();
+            while (current != null) {
+                final boolean[] isDeleted = {false};
+                @Nullable final Node next = current.next.get(isDeleted);
+                if (isDeleted[0]) {
+                    if (previous == null ? head.compareAndSet(current, next) : previous.next.compareAndSet(current, next, false, false)) {
+                        current = next;
+                        continue;
+                    } else {
+                        continue start;
+                    }
+                }
+                if (current.value.compareTo(value) >= 0) {
+                    break;
+                }
+                previous = current;
+                current = next;
+            }
+            return new PrevNextPair(previous, current);
+        }
+    }
+
+    @NotNull
+    private PrevNextPair findPrevNextRecover(@NotNull T value) {
+        return findPrevNext(value);
+    }
+
+    public boolean add(int p, @NotNull T value) {
+        checkPointer[p] = 0;
+        // flush, not necessary
+        @NotNull final Node newNode = new Node(value, null);
+        recoveryData.set(p, new Info(newNode));
+        // flush, not necessary
+        checkPointer[p] = 1;
+        // flush
+        while (true) {
+            @NotNull final PrevNextPair prevNext = findPrevNext(value);
+            @Nullable final Node previous = prevNext.previous;
+            @Nullable final Node next = prevNext.next;
+            if (next != null && next.value.compareTo(value) == 0) {
+                recoveryData.get(p).result = false;
+                // flush
+                return false;
+            }
+            newNode.next.set(next, false);
+            // flush
+            if (previous == null) {
+                if (head.compareAndSet(next, newNode)) {
+                    recoveryData.get(p).result = true;
+                    // flush
+                    return true;
+                }
+            } else {
+                if (previous.next.compareAndSet(next, newNode, false, false)) {
+                    recoveryData.get(p).result = true;
+                    // flush
+                    return true;
+                }
+            }
+        }
+    }
+
+    public boolean addRecover(int p, @NotNull T value) {
+        if (checkPointer[p] == 0) return add(p, value);
+        @Nullable Node node = recoveryData.get(p).node;
```

Comment:
Please don't mark local variable with `@Nullable` -- it is more complicated to follow the code logic with these annotations